fc/parsing/actions.py

```diff
@@ -1200,204 +1266,301 @@ def _number_generator(self, value, units):
+        To create a variable, it may be necessary to infer its units from an RHS specified by the user. This requires
+        the RHS to be available as a sympy expression where all variables have known units. Because the RHS can include
+        references to other variables yet to be created, these operations cannot be separated so must all be handled
+        here at once.
+        """
+        todo = set(self.protocol_variables.values())
+        while todo:
+
+            # A potential MissingVariableError encountered in this pass
+            error = None
+
+            done = set()
+            for pvar in todo:
+
+                # RHS to set in this iteration
+                rhs = None
+
+                # Add variable if required
+                if pvar.model_variable is None:
+
+                    # Check if there's enough information to define the variable's RHS
+                    if pvar.equation is None and pvar.initial_value is None and pvar.default_expr is None:
+                        # No definition given for variable, this is OK only if it's optional
+                        if pvar.optional:
+                            done.add(pvar)
+                            continue
+                        elif pvar.local:
+                            raise ProtocolError(f'No definition given for local variable {pvar.long_name}.')
+                        else:
+                            raise ProtocolError(f'No definition given for non-optional variable {pvar.long_name}.')
+
+                    # Get units to create variable with
+                    if pvar.units is not None:
+                        units = self.units.get_unit(pvar.units)
+                    else:
+                        # No units specified, try getting from RHS
+                        rhs = pvar.default_expr if pvar.equation is None else pvar.equation.rhs
+                        if rhs is None:
+                            # Not enough information to create, OK if optional, but otherwise an error
+                            if pvar.optional:
+                                done.add(pvar)
+                                continue
+                            else:
+                                raise ProtocolError(f'No units specified for non-optional variable {pvar.long_name}.')
+
+                        # Try getting sympy RHS
+                        # TODO: There's probably a faster way to check if we can resolve all variables in the RHS
+                        try:
+                            rhs = rhs.to_sympy(self._variable_generator, self._number_generator)
+                        except MissingVariableError as e:
+                            # Unable to create at this time, but may be able to at a next pass
+                            error = e
+                            continue
+
+                        # Set units based on RHS
+                        units = self.units.evaluate_units(rhs)
```

Comment:
This has come up again. Without having units here I can't create this variable

Reply:
I wasn't thinking straight when I made this change :D

`set_lhs_units_from_rhs` is not something we can do in cellmlmanip, is it? Because the variable units are fixed when the variable is created?

Reply:
So is the idea that I set them to e.g. dimensionless first, and then have the RHS override it? I think that would hide errors
Also the code in `set_lhs_units_from_rhs` doesn't deal with state variables, it seems

Reply:
I don't think we should have a `set_lhs_units_from_rhs`, really. It changes a variable's units without updating the model equations.

Reply:
I'll just deduce the correct units with `convert_expression_recursively` when I create the variable instead